semgrep-core/Parsing/other/yaml_to_generic.ml

```diff
@@ -0,0 +1,121 @@
+module Y = Yaml
+module S = Yaml.Stream
+module E = Yaml.Stream.Event
+module M = Yaml.Stream.Mark
+module A = AST_generic
+
+exception Unimplemented
+exception ParseError of string
+
+(* Parser for YAML. The tree-sitter grammar for YAML is somewhat complicated,
+ * since YAML is whitespace sensitive. Instead, we use the OCaml Yaml module,
+ * which we also use to parse YAML for semgrep. To get positions, we use the
+ * low-level Stream API, which we parse into a generic AST
+*)
+
+(* Helper functions *)
+
+let fst (a, _) = a
+
+let p_token = function
+  | E.Stream_start _ -> "stream start"
+  | E.Stream_end -> "stream end"
+  | E.Document_start _ -> "document start"
+  | E.Document_end _ -> "document end"
+  | E.Sequence_start _ -> "sequence start"
+  | E.Sequence_end -> "sequence end"
+  | E.Mapping_start _ -> "mapping start"
+  | E.Mapping_end -> "mapping end"
+  | E.Alias _ -> "alias"
+  | E.Scalar { value; _ } -> "scalar " ^ value
+  | E.Nothing -> "nothing"
+
+let mk_err err v {E.start_mark = {M.line; M.column; _}; _} =
+  Printf.sprintf "%s %s at line %d column %d" err (p_token v) line column
+
+let get_res = function
+  | Result.Error (`Msg s) -> raise (ParseError s)
+  | Result.Ok v -> v
+
+let any_of_list = function
+  | [x] -> A.E x
+  | xs -> A.E (A.Seq xs) (* Change to array? *)
+
+let tok (index, line, column) str file =
+  {Parse_info.token = Parse_info.FakeTokStr (str, Some ({str; charpos = index; line; column; file}, 0));
+   Parse_info.transfo = NoTransfo}
+
+let mk_tok {E.start_mark = {M.index; M.line; M.column}; _} str file =
+  tok (index, line, column) str file
+
+let mk_bracket ({E.start_mark = {M.index = s_index; M.line = s_line; M.column = s_column}; _},
+                {E.end_mark = {M.index = e_index; M.line = e_line; M.column = e_column}; _}) v file =
+  tok (s_index, s_line, s_column) "(" file, v, tok (e_index, e_line, e_column) ")" file
+
+(* Parser functions *)
+
+let make_alias anchor pos file : A.name =
+  A.Id ((anchor, (mk_tok pos "" file)), {id_resolved = ref None; id_type = ref None; id_constness = ref None})
+
+let make_scalar _anchor _tag pos value file : A.expr =
+  A.L (A.String (value, mk_tok pos value file))
```

Comment:
you can put a TODO here saying we may want to generate a A.Float or A.Bool depending on the string content.